### PR TITLE
Add methods to check native libraries

### LINKED ISSUE #175: Add a TryLoadingLibraryFiles method
* **I'm submitting a ... **
  - [ ] bug report
  - [X] feature request
  - [ ] support request or question => Please do not submit support request or questions here, see note at the top of this template.

* **What is the current behavior?***
For windows and Linux and maybe MacOS there can be different folders where to lookout for the library files (Current folder, PATH system environment variable, fixed path under Linux, sub folder, etc.pp). Currently we can either enumerate the files ourselves on all of these places before setting the RootPath property (but `ffmpeg.autogen` does not provide the full filenames directly to us) or simply set it and hope no DllNotFound exception raises. See also https://github.com/Ruslan-B/FFmpeg.AutoGen/blob/master/FFmpeg.AutoGen.Example/FFmpegBinariesHelper.cs there is no property validation that tells us if the library files really exist or not.

* **What is the expected behavior?**
I suggest adding a simple helper method that returns a boolean if all library files exist. Preferably if all library files exist they also get loaded:

```C#
public bool TryLoadingLibraryFiles(string? path = null) {
	path ??= RootPath;
	bool allFound = false;
	foreach (var item in LibraryDependenciesMap) {
		//if File.Exists(...)
	}
	return allFound;
}
```

* **What is the motivation / use case for changing the behavior?**
We can conveniently write something like this to look at multiple paths:
```C#
string rootPath = ffmpeg.RootPath;
if (!ffmpeg.TryLoadingLibraryFiles()) {
	if (!ffmpeg.TryLoadingLibraryFiles(Path.Combine(rootPath, "bin"))) {
		ffmpeg.RootPath = Path.Combine(rootPath, "ffmpeg");
		//....
	}
}
```

* **Please tell us about your environment:**

- version: 4.3.0.3

### PR DESCRIPTION
**Pull request by Speykious**
This PR aims at making it easier to _check_ if native libraries can be loaded given the current root path.
Closes #175, though it had been technically closed due to PR #181 not getting merged.

Unlike #181, this PR doesn't try to load or unload any native libraries, but simply checks if the file is there.
It also checks for library dependencies the same way the `LoadLibrary()` method does.
Finally, like previously, I also give a way to specify which libraries are required, so that it can check them alone without the other ones.

Now that I think about it, this implementation does indeed look way less hacky than the previous one. I think I was just unsure as to whether checking the files would be enough; but then I realized that it loads the libraries from their full path, so it won't cause any problem.

Has been successfully tested on Linux once again, but not on Windows, though the logic should be the same.
I modified the `FFmpegBinariesHelper` class in the example as well to show how the functions would be used.

### CODE REVIEW

File: FFmpeg.AutoGen.Example/FFmpegBinariesHelper.cs [hunk 1]
```diff
@@ -10,27 +10,30 @@ internal static void RegisterFFmpegBinaries()
...
                     current = Directory.GetParent(current)?.FullName;
                 }
             }
             else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
-                ffmpeg.RootPath = "/lib/x86_64-linux-gnu/";
+            {
+                ffmpeg.TrySetRootPath(
+                    LibraryFlags.AVCodec,
+                    AppDomain.CurrentDomain.BaseDirectory,
+                    "/usr/lib",
```
Comment (kant2002) [inline FFmpeg.AutoGen.Example/FFmpegBinariesHelper.cs:29]: On Gentoo, for example this path is `/usr/lib64`

File: FFmpeg.AutoGen/FFmpeg.cs [hunk 2]
```diff
@@ -51,9 +52,78 @@ static ffmpeg()
...
+        public static bool CanLoadLibraries(LibraryFlags libraries = LibraryFlags.All, string path = "")
+        {
+            var validated = new List<string>();
+            return libraries.ToStrings().All((lib) => canLoadLibrary(lib, path, validated));
+        }
+
+        // Note: recurses in a very similar way to the LoadLibrary() method.
+        private static bool canLoadLibrary(string lib, string path, List<string> validated)
+        {
+            Console.WriteLine($"Checking if {lib} can be loaded from {path}");
```
Comment (kant2002) [inline FFmpeg.AutoGen/FFmpeg.cs:73]: I would like to have this kind of tracing when something goes wrong, but I don't want that noise in my application if everything is ok.

**Comment by Ruslan-B** [comment]
I am not going to approve this PR any way as no rationale to pull all this stuff into library.
It can be done as dedicated and function in the user application - as you have access to all metadata (like libraries and versions) and to ```GetNativeLibraryName```.
I guess I have to amend test project with the example.

**Comment by Speykious** [comment]
Again? Even after I removed all the need for even loading any library in the first place? Why? This time I just don't understand.

The rationale is simple: _all_ people who use FFmpeg.AutoGen want a way to easily check for libraries available in the current root path. Asking them to write all this code by themselves is a lighter version of asking them to write the code to load the native libraries themselves with the corresponding OS functions so that the bindings could work.
Anything else that I have said about our current solutions has been said in the closed issue on the topic.

That I overlooked a `Console.WriteLine()` which I forgot to remove while debugging (since I use both prints and the debugger to debug stuff) is something completely different and that I can remove immediately upon change requests.

Please, can you tell me why you think that it should be better to let everyone else write this code instead of providing it directly in FFmpeg.AutoGen?

File: FFmpeg.AutoGen/FFmpeg.cs (hunk 2 above)
Comment (Speykious) [reply FFmpeg.AutoGen/FFmpeg.cs:73]: Yes, sorry, that was a statement that I wrote for debugging and forgot to remove.

**Comment by kant2002** [comment]
@Speykious may I try to explain why @Ruslan-B is not willing to take that specific contribution? Please consider, that I do not know him, and this is my speculation based on my maintenance experience on other projects.

The difference between your approach and Ruslan's lays in the perception what people this library should server. Ruslan thinks, this library is a tool which each developer use to connect to ffmpeg libraries.You think, that this is library which helps you write user facing applications. At this point I should be corrected if I'm wrong.

Both approaches has different philosophy if you like. If you wrote tool, you try to make it as plain as possible, so trained developer would not think too much about a tool. If you write library which helps developers then you focus on shielding developers from the underlying tech, in order to make library easy to use. Both points contradicting, but not mutually exclusive.

I would speculate, that Ruslan wrote tool which provide PInvoke methods for all ffmpeg libraries and stops at this point. Same as .NET which provide `DllImport` and a bit of other controls. It's developer responsibility to place libraries where they belongs.

If you think about library which provide access for FFMpeg , then you are right, you want smooth way to find FFMPEG binaries. then this kind of logic is appropriate. If you want look at library similar to your taste, you may look at https://github.com/radek-k/FFMediaToolkit/ where maintainers try to make it work according to your tastes.

If you notice, this library do not try to hide names of FFmpeg functions from you, even to C# developers they are too hard to read. That's deliberate decision, so C/C++ samples can be used as reference material.

Would you consider following scenarios which make you claim `all people who use FFmpeg.AutoGen want a way to easily check for libraries available in the current root path` weak.

Scenario 1
Desktop application for Windows which package ffmpeg binaries into installer, and place them in the root of the application. Developer do not need touch `RootPath` at all.

Scenario 2
Server application on Linux for enterprise. App load value of `RootPath` from configuration file. It is responsibility of client's sysadmin to properly configure your application for their own distro. Nothing for this library to do.

Would you try to explain what scenario require advanced multi-path probing?

**Comment by Speykious** [comment]
> You think, that this is library which helps you write user facing applications. At this point I should be corrected if I'm wrong.

Indeed. I don't think it helps doing that. I understand very well that FFmpeg.AutoGen is meant to be nothing more than a 1-to-1 C# port of the C APIs. It is actually the reason I use it, because I am familiar with the C APIs. Providing any additional FFmpeg-related abstraction would be the job of a wrapper library, like [FFMediaToolkit](https://github.com/radek-k/FFMediaToolkit) or [EmguFFmpeg](https://github.com/IOL0ol1/EmguFFmpeg).

> I would speculate, that Ruslan wrote tool which provide PInvoke methods for all ffmpeg libraries and stops at this point. Same as .NET which provide `DllImport` and a bit of other controls. It's developer responsibility to place libraries where they belongs.

And it will still be the case if my PR was to be accepted. All my PR is doing is making it easier for users of FFmpeg.AutoGen, whether it be developers of wrappers mentioned above or of applications like [osu!](https://github.com/ppy/osu), to set that path. It won't make the application check for paths automatically, because as I explained previously in the closed issue, that will still be the responsibility of the application which goals will make that use-case different.

***

> Would you try to explain what scenario require advanced multi-path probing?

Well, first-off, I wouldn't consider this "advanced". All `TrySetRootPath()` is doing is checking if the libraries are available in the different paths given as parameters, and setting `RootPath` if it manages to find them. Not to mention that:
- I designed it so that it doesn't assume what libraries the user needs, because of all possible use-cases. For example, the first scenario you mentioned might be a simple video encoder and thus would only need `avcodec` and `avformat`. The second scenario might be streaming some downscaled webcam output to the internet and would thus also require `avdevice` and `swscale`.
- Even if these scenarios only require one path, using `TrySetRootPath` might throw an exception at a more appropriate time if the libraries are not detected, rather than at the first FFmpeg PInvoke method that needs to have the runtime loaded (though I would need to improve this PR for that since right now it doesn't throw a meaningful exception).
- Doing `ffmpeg.TrySetRootPath(thePath);` instead of `ffmpeg.RootPath = thePath;` changes pretty much nothing code-wise, unless you want to specify the actually required libraries, in which case it will only add a few characters, and it adds the advantage that I mentioned in my previous point.

Again, the only goal of this PR is to make it easier to let the users of FFmpeg.AutoGen find the FFmpeg native libraries. I don't believe that it needs to be the responsibility of the users to provide such help. To _actually find them_ would be.

**Comment by Ruslan-B** [comment]
@kant2002 Thank you for the explanation. 
My personal view: dll importing and function resolution already quite complex process. This PR bumps complexity even more. So I am inclined to apply Occam's razor here.
Simple pre check function can do all the work - before setting up ```RootPath```. All necessary metadata and functions to form platform specific naming are publicly available. 
If you need an example I can provide one

**Comment by Speykious** [comment]
I am already doing so [here](https://github.com/vignetteapp/SeeShark/blob/main/SeeShark.Example.Ascii/FFmpegLoader.cs), since this PR got rejected.

I still think that this function would help everyone as it is exactly a _simple pre-check function_. As such, I don't think it bumps the complexity in any manner.

Maybe I can remove the enum that I added so that, just like in my example above, it checks directly for the names of the libraries instead, if you see that enum as unnecessary, but still.

**Comment by Speykious** [comment]
I will also mention that if some use-case only requires one library, then assuming that it is the responsibility of the user, they would have to check for the dependencies as well (which is what I'm doing in `CanLoadLibraries`). So they would have to write the recursive function themselves.

It's not like it's the end of the world of course, but in my view, since all they're doing is checking that libraries are present so that they can set a correct path, it is definitely something that `FFmpeg.AutoGen` can provide. It already has a dependency and version map, it already has functions to get the names of the libraries per platform; this is honestly the last thing it needs in this area (again in my own opinion).

**Comment by Ruslan-B** [comment]
@Speykious I got you point - however, we all have our styles and preferences, I'll try to resolve this discussion to tangible code in the weekend.

**Pull request closed without merging**